src/InvoiceManager.php

```diff
@@ -0,0 +1,326 @@
+        if (null != $first_invoice->getMeta() || null != $second_invoice->getMeta()) {
+            $invoice->meta = array_merge($first_invoice->getMeta() ?? [], $second_invoice->getMeta() ?? []);
+        }
+        $invoice->amount = $first_invoice->amount->getValue() + $second_invoice->amount->getValue();
+        if (isset($localizedDetails['title'])) {
+            $invoice->title = $localizedDetails['title'];
+        }
+        $invoice->save();
+        $first_invoice->products()
+                      ->update([
+                                   'invoice_id' => $invoice->getID(),
+                               ]);
+        $second_invoice->products()
+                       ->update([
+                                    'invoice_id' => $invoice->getID(),
+                                ]);
+
+        return $invoice;
+    }
+
+    public function addPaymentToInvoice(int $invoiceId, string $type, INumber $amount, PaymentStatus $status, ?array $meta): IPayment
+    {
+        // TODO: Implement addPaymentToInvoice() method.
+        $invoice = $this->getInvoiceById($invoiceId);
+        if (InvoiceStatus::PAID == $invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        $payment = $invoice->payments()
+                           ->create([
+                                        'method' => $type,
+                                        'amount' => $amount,
+                                        'meta' => $meta,
+                                        'status' => $status,
+                                    ]);
+
+        return $payment;
+    }
+
+    public function approvePayment(int $paymentId): IPayment
+    {
+        // TODO: Implement approvePayment() method.
+        $payment = Payment::query()
+                          ->findOrFail($paymentId);
+        if (InvoiceStatus::PAID == $payment->invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        if ($payment->getAmount() > $payment->invoice->getAmount()) {
+            throw new AmountInvoiceMismatchException();
+        }
+        if ($payment->getAmount() == $payment->invoice->getAmount()) {
+            $payment->update([
+                                 'status' => PaymentStatus::APPROVED,
+                             ]);
+            $payment->invoice->update([
+                                          'status' => InvoiceStatus::PAID,
+                                          'paid_amount' => $payment->getAmount(),
+                                      ]);
+        }
+
+        return $payment;
+    }
+
+    public function rejectPayment(int $paymentId): IPayment
+    {
+        // TODO: Implement rejectPayment() method.
+        $payment = Payment::query()
+                          ->findOrFail($paymentId);
+        if (InvoiceStatus::PAID == $payment->invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        $payment->update([
```

Comment:
What if payment already approved? no need to check status of payment?

Reply:
done